### ISSUE TITLE
md-content doesn't work correctly with flex.

### ISSUE THREAD

**Issue by langley-agm** [angular/material#4022]
Maybe it works as expected but if it does, I'm not sure I fully understand the purpose of md-content and the documentation resumes up to one single line. I made a shorter example of my issue.

I want two divs with overflowing content each covering half the page, 

This example works as I would expect (the two divs are immediate children of the body):

http://codepen.io/langley/pen/Nqorzv

However if they are not direct children of the body, but a div instead, or even another md-card, md-whiteframe or any other, it stops working:

http://codepen.io/langley/pen/waNWYe

In order to make it work I have to apply height: inherit or height: 100% to the parent:

http://codepen.io/langley/pen/KpJMbR

It's not a big deal in this example but if the object I want to be scrollable is a nested child of 10 elements, I have to apply this to every single parent, not just the closest. 

I see that in the demo a fixed px value is used, but I thought the whole idea of being responsive is to use flex-xx style and % instead of fixed px values ?

**Comment by nkoterba** [comment]
@langley-agm I'm not sure how to do that cool little thumbs up emoticon, so I'll just have to do:

+1 +1 +1

Been struggling all day to get a `md-content` nested inside several layers of `divs` with `layout="column"` and "flex" to correctly scroll.

Using latest chrome and MD `0.10.0` (no RCs yet for me).

I agree that the demo of `md-content` uses a fixed height, which always works but like you said totally defeats the purpose of flexbox.

I'm not sure if this is a bug, intentional design, or what.

My current **hack/workaround** is to give the md-content a style of `height: 1px`.  This seems to trick (at least Chrome, nervous about FF and IE) into correctly "boxing" `md-content` so my inner "overflowing" content scrolls (and not the whole parent div).

http://codepen.io/anon/pen/MwLQqJ

Here's another code-pen that is the current demo MD uses in their documentation for `md-content`.  All I've done is add more content to ensure scrolling will occur and remove the "fixed" height of the `md-content` element.

You see the same behavior you describe above.

What I would expect to happen is that if I add `flex` to the `md-content`, it should fill all the space after the `md-toolbar` but stop at the bottom of the parent `div` and the `md-toolbar` should **not** scroll, just the content in the `md-content` container.

**Comment by juandiago** [comment]
@nkoterba thnks for the <code>height:1px</code> workaround.

Works like a charm!

**Comment by topherfangio** [comment]
Hi all, I am looking into this issue, but I'm having trouble understanding the `height: 1px` fix/workaround. When I apply that to the `md-content`, it gives me a `1px` tall block of text (which doesn't seem like a proper fix).

Can one of you produce a quick demo that shows the issue and has the CSS fix comment out so that I can comment it back in and see the difference?

**Comment by juandiago** [comment]
Here you go:

http://codepen.io/anon/pen/gpyEjV

Just remove the <code>height:1px</code> from the <code>md-content</code>

Hope it helps!..

**Comment by nkoterba** [comment]
@topherfangio @juandiago demo is a good example.  I updated the codepen to use the "master" CSS and JS files as it was still on 0.10 (instead of 0.10.1 or master):

http://codepen.io/anon/pen/vOMMbX

Issue still exists.  As @juandiago said, if you remove the `height: 1px` then the entire "page" scrolls versus just the stuff in the top `md-content`.

**Comment by topherfangio** [comment]
I can confirm that the following code should fix the issue:

```css
md-content[flex] {
  height: 0px;
}
```

We have a team meeting today, let me confirm with the others that this shouldn't break anything else.

**Comment by topherfangio** [comment]
Seems the same as #3901. We're still trying to determine if there is a better fix or not.

**Comment by ThomasBurleson** [comment]
Closed with SHA 303ab0dc0e695f

**Comment by topherfangio** [comment]
Due to some unexpected breaking in other developers applications (#4397), we have decided to remove this change from 0.11 and get some more community feedback before pushing a fix in 0.12.

Reverted in commit 44a7e52c5832e9e8ee31d9c67de1bc4def77bf22

**Comment by nkoterba** [comment]
Just wanted to say this is **definitely** still broken as of Sept 28, 2015 on the `HEAD` version.

Check out the codepen: https://codepen.io/anon/pen/avBMKd

Thankfully, I followed #4453, which suggests the following fix that I **currently have commented out** in the codepen.

Just re-enable the suggested "fix" CSS, and md-content scrolls again!

**Comment by ThomasBurleson** [comment]
The latest stability fixes in `layout.css` **and** tweaks to your demo fixes this:

https://codepen.io/ThomasBurleson/pen/MaJwqV

> Note the overrides to the `docs.css` with:
```css
body {
  height : 100%;
  overflow : auto; 
}
```

**Comment by ThomasBurleson** [comment]
@nkoterba - thx for the alert. I think the [Content demo](http://localhost:8080/#/demo/material.components.content) should be updated also.

**Comment by langley-agm** [comment]
I still see the issue in:

https://codepen.io/ThomasBurleson/pen/MaJwqV

using chrome, the scrollbar is also scrolling the toolbar and not just the md-content.

**Comment by langley-agm** [comment]
btw the Content Demo you linked in the previous comment is pointing to localhost.

**Comment by ThomasBurleson** [comment]
@langley-agm - I will investigate.

**Comment by ThomasBurleson** [comment]
Fixed both:  https://codepen.io/ThomasBurleson/pen/MaJwqV

**Comment by langley-agm** [comment]
Sorry, I might not be explaining myself correctly, I updated the initial examples to use master instead. 
They behave exactly as they were behaving when I raised this issue.

Manually applying height: 100% or inherit (which inherits the 100% from the body) has allways fixed this issue, which is what you are doing in the example above. but if the component is too nested you'd have to go over each parent doing this, is this the expected thing to do? if so, what's the purpose of md-content component?

**Comment by topherfangio** [comment]
@langley-agm I have been working on a potential fix for this issue. There are two pieces to it:

1. You need to set the width/height of the layout container to fill all available space, and
2. The closest `flex` parent of the `md-content` needs to have `overflow: auto` set.

For #1, you can simply use the `layout-fill` directive; however, if your beginning layout is nested within a few `<div>` tags, they might also need `layout-fill`. One potential solution we are investigating is to add a `layout-root` directive that sets the following CSS which alleviates the issue:

```css
position: absolute;
left: 0;
right: 0;
top: 0;
bottom: 0;
```

For #2, we're looking at updating the `md-content` directive to do this automatically.

Here is a Codepen which shows a basic version of both of these solutions:http://codepen.io/topherfangio/pen/meMNVJ

Do you feel like there is any merit to #1 above? If your main app layout isn't too nested, then adding the `layout-fill` to an extra div or two doesn't seem like much work, but if it's deeply nested, then I can definitely see the need for the `layout-root` directive.

**Comment by langley-agm** [comment]
Hi @topherfangio , I wouln't mind having to repeat layout-fill, that's one of the approaches I tried but couldn't make it work. 
Just to be clear though, what would I have to do to make this example:

https://codepen.io/langley/pen/RWLEda

look like this: http://codepen.io/langley/pen/Nqorzv ?

**Comment by topherfangio** [comment]
@langley-agm So, in your case, the proposed `layout-root` directive really wouldn't help, but the `layout-fill` definitely would.

Here is your first Codepen modified to look like the second: https://codepen.io/topherfangio/pen/MaExNo

Note that the extra CSS is only there to fix Firefox not properly calculating the height. I was hoping that the `md-content` directive could do this, but it looks like it can't easily solve every case, so I think this is the best approach at the moment.

Let me know if this works for you!

**Comment by langley-agm** [comment]
Seems clean and simple enough, I'm not so fun of the required firefox fix since its a really simple use case, hopefully it can be specified in the documentation, but for now that's the way I'll go for my app, thank you !

**Comment by topherfangio** [comment]
@langley-agm Glad it worked! I agree that the CSS fix for Firefox is a bit annoying, but unfortunately we haven't found a clean way to workaround it at a "global" level without messing up other applications.

We'll continue to look for a better fix as we push towards the 1.0 release.

**Comment by derekmoss** [comment]
I modified the codepen you just posted and just deleted some of the 'stuff' from the first scroll area only

https://codepen.io/jayray/pen/mexQZq

each row in the column no longer divides the space equally.
When the scroll extent is larger in one than the other the md-contents ignore flex.

I posted a issue yesterday, It seems related.

https://github.com/angular/material/issues/5340

and for quick reference the codepen i made to demonstrate the problem
http://codepen.io/jayray/pen/yYvwMZ
